### ISSUE TITLE
Why clamp is needed during training

### ISSUE THREAD

**Issue by bluesky314** [PeterL1n/BackgroundMattingV2#128]
Why do we need to clamp alpha and fg between 0,1 during training? We could just clamp it during inference

**Comment by bluesky314** [comment]
Why would you do so given how the backward is implemented which kills of the gradient: https://github.com/pytorch/pytorch/blob/53fe804322640653d2dddaed394838b868ce9a26/torch/autograd/_functions/pointwise.py#L95

**Comment by PeterL1n** [comment]
It performs better with a clamp. It prevents penalizing the model from not predicting exactly 0 or 1 for most areas, which is hard. It is a compromise, even if using sigmoid, the gradient will be close to zero when saturating at 0 or 1.

**Comment by PeterL1n** [comment]
It is particularly problematic when evaluating against metrics. If no clamping, the model has a hard time predicting exact 0, if it outputs 0.01, it will be measured negatively against the metric. If use a clamp, the network can easily overshoot, to make sure the area has exact 0, to minimize evaluation loss.

**Comment by bluesky314** [comment]
For metric evaluation, clamping is justified as it is post processing  step. However during training, if the network predicts a greater than 1 value for an alpha of 0.7 then there will be no gradient due to the clamping. This could even happen frequently at the start of training. What do you say to this?

**Comment by PeterL1n** [comment]
All I can say is that clamping works. It is not a main part of the paper, and you are free to experiment with whatever you like in your own implementation.

**Comment by bluesky314** [comment]
Ok thanks, understood about the clamping. Any paticular reason you did not go with the conventional sigmoid activation function?

**Comment by PeterL1n** [comment]
FBA paper has more explanation.

Sigmoid is more suitable for regular RGB images where most of the values are not straight 0 or 1. Using sigmoid for alpha pushes the logits to very big in order to satisfy 0 and 1 outputs. Still, there are times where the network outputs 0.999, and if you use int(out * 255), it usually gets converted to 254, which hurts metrics.
